docs/plots/backends.rst

```diff
@@ -0,0 +1,362 @@
+************************
+Plotting backend details
+************************
+
+.. warning::
+   The plotting backend API is currently re-designed and details might change in the near future.
+
+Plotting in Sherpa is done through *plotting backends*. A *backend* is an
+external package that performs the actual plotting function such as :term:`matplotlib`, i.e. it creates a
+canvas and puts color on paper or on the screen. The Sherpa plotting classes are
+written in a backend-independent way; they work with any plotting backend that
+is supported by Sherpa. We also use the term *backend* to refer to Sherpa
+classes that connect the Sherpa plot objects with the plotting backend by
+translating Sherpa options into backend-specific commands. The details of that
+are explained below, but, in short, a backend class would translate the sherpa
+backend-independent "plot a line in red from here to there" into backend
+specific ``plt.plot(x,y, linecolor='r')`` (if the backend is :term:`matplotlib`).
+
+Instead of utilizing Sherpa classes and commands to plot data, users can
+alternatively just access the data in the Sherpa data objects (e.g. the x and y
+values of a dataset) and perform plotting operations directly with any plotting
+backend that is available to them. This method may be less convenient, but it
+works for ploting backends not (yet) supported by Sherpa.
+
+Which backend is used?
+======================
+
+When the `sherpa.plot` module is first imported, Sherpa tries to import the
+backends installed with Sherpa in the order listed in the
+``options.plot_pkg`` setting from the ``sherpa.rc`` startup file.
+The first module that imports successfully is set as the active
+backend. The following command prints the name and the location
+on disk of that module::
+
+   >>> from sherpa import plot
+   >>> print(plot.backend)
+
+Change the backend
+------------------
+
+.. warning::
+   When the plotting backend is changed, new plotting objects must be
+   initialized because objects like e.g. `sherpa.plot.DataPlot` keep a
+   copy of all default parameters of the backend that was active when they
+   are initialized. A different backend might not understand all those
+   parameters.
+
+   In particular, that chnging the backend might not work for plot functions
+   in Sherpa's UI layer such as `sherpa.astro.plot.plot_data` because those
+   work with objects underneath that are not updated when the plotting backend
+   is changed.
+
+After the initial import, the backend can be changed by loading one of
+the plotting backends shipped with sherpa (or any other module that
+provides the same interface)::
+
+  >>> import sherpa.plot.pylab_backend
+  >>> plot.backend = sherpa.plot.pylab_backend
+
```

Comment:
This doesn't seem to match now. I have matplotlib installed but the plot_pkg is set to `dummy`:

```
$ grep '^plot_pkg' ~/.sherpa-standalone.rc
plot_pkg   : dummy
```

which gives this

```
In [1]: from sherpa import plot

In [2]: print(plot.backend)
<sherpa.plot.backends.BasicBackend object at 0x7f899740c460>

In [3]: import sherpa.plot.pylab_backend

In [4]: plot.backend = sherpa.plot.pylab_backend

In [5]: print(plot.backend)
<module 'sherpa.plot.pylab_backend' from '/home/dburke/miniconda3/envs/sherpa-pr/lib/python3.10/site-packages/sherpa/plot/pylab_backend.py'>
```

Note that this is now referring to a module, not a backend class.

If I switch to picking up pylab automatically we get the class (as shown above but included here):

```
% grep '^plot_pkg' ~/.sherpa-standalone.rc
plot_pkg   : pylab
```

and

```
In [1]: from sherpa import plot

In [2]: print(plot.backend)
<sherpa.plot.pylab_backend.PylabBackend object at 0x7fefb981db10>
```

Reply:
I assume this should read something like the following (which will mean that there should be some docs to indicate that backend class to use for a module, although you can find it pretty easily):

```
sherpa.plot.backend = sherpa.plot.pylab_backend.PylabBackend()
```

Reply:
I've been able to have the dummy plot backend configured, start a notebook and have no plots (e.g. plot_data warns me no plot and get_data_plot() returns a table) and then switch to the pylb backend with the above and then new plots work and get_data_plot now contains a svg plot rather than a table.

I think we should look at making it easier to change the backend - have some routine that you can send the setting in as if you were editing the config file and have the system do the rest - but that would be in a follow-on PR.

Reply:
Sorry, yes, that's old. I do have a local branch to run doctests on the rst files that would have caught this, but there are other things with that branc hthat don't work yet (which is why I don't have a PR with it yet). 

I'll ponder about other ways to change the backend. This is the kind of comment I was expecting; I honestly had not checked yet if the references in the sphinx docs are correct, because I expected references on a higher level. There would be another way to implement that same idea for backends: They could be plugins: https://packaging.python.org/en/latest/guides/creating-and-discovering-plugins/
That would actually be a natural way to do this - but I'm I afraid using setuptools or pkg_utils as suggested in the link above will cause further difficulties for our ongoing changes to the installtion process. Maybe I can revisit that or find some middle ground to all backends by their names as opposed to the fully qualified path.